### PR TITLE
Try Nesting Tests Inside Package

### PR DESCRIPTION
**Pull request by mitchute**

### CODE REVIEW

**Comment by mitchute** [comment]
@JingWang-CUB ignore the ruff failures for now. Can you try this to see if it fixes your testing issues?

@vtnate if this works, maybe we can polish this up and use it instead of #33 ?

**Comment by vtnate** [comment]
> @vtnate if this works, maybe we can polish this up and use it instead of #33 ?

@mitchute This works on my mac. Needs a little cleanup that I can take care of.
@JingWang-CUB, Will you pull this version and try running the tests in a Python 3.9 environment? If it works, that would mean we don't have to change the python requirement.

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -8,7 +8,7 @@ jobs:
     strategy:
       matrix:
         python-version: ['3.9', '3.12']
-        os: ['ubuntu-latest', 'windows-latest']
+        os: ['ubuntu-latest', 'macos-latest', 'windows-latest']
```
Comment (vtnate) [inline .github/workflows/tests.yml:11]: I'm not sure this tells us any additional information beyond running on ubuntu.

File: thermalnetwork/ground_heat_exchanger.py [hunk 2]
```diff
@@ -101,6 +101,6 @@ def ghe_size(self, total_space_loads, output_path: Path) -> float:
     def get_atlanta_loads(self) -> list[float]:
         # read in the csv file and convert the loads to a list of length 8760
         current_file_directory = Path(os.path.dirname(os.path.abspath(__file__)))
-        glhe_json_data = current_file_directory / "Atlanta_loads/Atlanta_Office_Building_Loads.csv"
+        glhe_json_data = current_file_directory / "test_data" / "Atlanta_Office_Building_Loads.csv"
```
Comment (vtnate) [inline thermalnetwork/ground_heat_exchanger.py:104]: ❤️

File: thermalnetwork/tests/test_base.py [hunk 3]
```diff
@@ -52,7 +52,7 @@ def setUp(self) -> None:
         self.test_outputs_path.mkdir(exist_ok=True)
 
         # Save the original borehole length and number of boreholes, so they can be restored after the tests run.
-        sys_param_1_ghe = json.loads((self.system_parameter_path_1_ghe).read_text())
+        sys_param_1_ghe = json.loads(self.system_parameter_path_1_ghe.read_text())
```
Comment (vtnate) [inline thermalnetwork/tests/test_base.py:55]: We need to read the text first or else `loads` gets whiny

**Review by vtnate — APPROVED** [review]
This works on Mac with Python 3.12. If it also works on Windows then we're hunky dory!

File: thermalnetwork/tests/test_base.py (hunk 3 above)
Comment (mitchute) [reply thermalnetwork/tests/test_base.py:55]: Isn't that what is happening here?

File: .github/workflows/tests.yml (hunk 1 above)
Comment (mitchute) [reply .github/workflows/tests.yml:11]: Yeah, that's probably true. Maybe we should also try the `commadore64-latest` runner  💻

**Review by mitchute — COMMENTED** [review]
This is an artifact of the differences between PyCharm's native formatter and Ruff. Maybe I can configure PyCharm to pick up a Ruff formatting file...

**Comment by JingWang-CUB** [comment]
**_I'm still having test failures on Windows with Python 3.9. See errors below._** 
============================= test session starts =============================
platform win32 -- Python 3.9.0, pytest-7.4.4, pluggy-1.5.0
rootdir: C:\GitHub\ThermalNetwork
configfile: pyproject.toml
testpaths: thermalnetwork/tests
plugins: cov-4.1.0, syrupy-4.6.1
collected 0 items / 3 errors

=================================== ERRORS ====================================
_____ ERROR collecting thermalnetwork/tests/test_ground_heat_exchanger.py _____
ImportError while importing test module 'C:\GitHub\ThermalNetwork\thermalnetwork\tests\test_ground_heat_exchanger.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
C:\Users\jwang5\Anaconda3\envs\gmt\lib\importlib\__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
thermalnetwork\tests\test_ground_heat_exchanger.py:1: in <module>
    from thermalnetwork.ground_heat_exchanger import GHE
E   ModuleNotFoundError: No module named 'thermalnetwork'
___________ ERROR collecting thermalnetwork/tests/test_heat_pump.py ___________
ImportError while importing test module 'C:\GitHub\ThermalNetwork\thermalnetwork\tests\test_heat_pump.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
C:\Users\jwang5\Anaconda3\envs\gmt\lib\importlib\__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
thermalnetwork\tests\test_heat_pump.py:3: in <module>
    from thermalnetwork.heat_pump import HeatPump
E   ModuleNotFoundError: No module named 'thermalnetwork'
____________ ERROR collecting thermalnetwork/tests/test_network.py ____________
ImportError while importing test module 'C:\GitHub\ThermalNetwork\thermalnetwork\tests\test_network.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
C:\Users\jwang5\Anaconda3\envs\gmt\lib\importlib\__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
thermalnetwork\tests\test_network.py:5: in <module>
    from thermalnetwork.network import run_sizer_from_cli_worker
E   ModuleNotFoundError: No module named 'thermalnetwork'

----------- coverage: platform win32, python 3.9.0-final-0 -----------
Coverage HTML written to dir htmlcov

=========================== short test summary info ===========================
ERROR thermalnetwork/tests/test_ground_heat_exchanger.py
ERROR thermalnetwork/tests/test_heat_pump.py
ERROR thermalnetwork/tests/test_network.py
!!!!!!!!!!!!!!!!!!! Interrupted: 3 errors during collection !!!!!!!!!!!!!!!!!!!
============================== 3 errors in 0.22s ==============================

**_All tests pass on Windows with Python 3.10._**
============================= test session starts =============================
platform win32 -- Python 3.10.12, pytest-7.4.4, pluggy-1.3.0
rootdir: C:\GitHub\ThermalNetwork
configfile: pyproject.toml
testpaths: thermalnetwork/tests
plugins: anyio-4.4.0, cov-4.1.0, syrupy-4.5.0
collected 6 items

thermalnetwork\tests\test_ground_heat_exchanger.py .                     [ 16%]
thermalnetwork\tests\test_heat_pump.py .                                 [ 33%]
thermalnetwork\tests\test_network.py ....                                [100%]

---------- coverage: platform win32, python 3.10.12-final-0 ----------
Coverage HTML written to dir htmlcov


======================== 6 passed in 221.62s (0:03:41) ========================

**Comment by mitchute** [comment]
I sometimes have this issue when using VSCode (on Mac) because... (IDK... the env vars don't get set properly??) What I usually have to do is add the following.

`$ PYTHONPATH=. pytest` ... 

i.e. the "PYTHONPATH" part. Does that help?

**Comment by JingWang-CUB** [comment]
> I sometimes have this issue when using VSCode (on Mac) because... (IDK... the env vars don't get set properly??) What I usually have to do is add the following.
> 
> `$ PYTHONPATH=. pytest` ...
> 
> i.e. the "PYTHONPATH" part. Does that help?

It might have helped a little. Now I got different errors. 
$ PYTHONPATH=. pytest
============================= test session starts =============================
platform win32 -- Python 3.9.0, pytest-7.4.4, pluggy-1.5.0
rootdir: C:\GitHub\ThermalNetwork
configfile: pyproject.toml
testpaths: thermalnetwork/tests
plugins: cov-4.1.0, syrupy-4.6.1
collected 1 item / 2 errors

=================================== ERRORS ====================================
_____ ERROR collecting thermalnetwork/tests/test_ground_heat_exchanger.py _____
ImportError while importing test module 'C:\GitHub\ThermalNetwork\thermalnetwork\tests\test_ground_heat_exchanger.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
C:\Users\jwang5\Anaconda3\envs\gmt\lib\importlib\__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
thermalnetwork\tests\test_ground_heat_exchanger.py:1: in <module>
    from thermalnetwork.ground_heat_exchanger import GHE
thermalnetwork\ground_heat_exchanger.py:7: in <module>
    from ghedesigner.manager import GHEManager
E   ModuleNotFoundError: No module named 'ghedesigner'
____________ ERROR collecting thermalnetwork/tests/test_network.py ____________
ImportError while importing test module 'C:\GitHub\ThermalNetwork\thermalnetwork\tests\test_network.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
C:\Users\jwang5\Anaconda3\envs\gmt\lib\importlib\__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
thermalnetwork\tests\test_network.py:5: in <module>
    from thermalnetwork.network import run_sizer_from_cli_worker
thermalnetwork\network.py:10: in <module>
    from rich.logging import RichHandler
E   ModuleNotFoundError: No module named 'rich'

----------- coverage: platform win32, python 3.9.0-final-0 -----------
Coverage HTML written to dir htmlcov

=========================== short test summary info ===========================
ERROR thermalnetwork/tests/test_ground_heat_exchanger.py
ERROR thermalnetwork/tests/test_network.py
!!!!!!!!!!!!!!!!!!! Interrupted: 2 errors during collection !!!!!!!!!!!!!!!!!!!
============================== 2 errors in 0.66s ==============================

**Comment by mitchute** [comment]
`pip freeze` shows ghedesigner and rich?

File: thermalnetwork/tests/test_base.py (hunk 3 above)
Comment (vtnate) [reply thermalnetwork/tests/test_base.py:55]: Sorry, I wasn't clear. Yes, the way I have it does exactly that.

**Comment by vtnate** [comment]
> This is an artifact of the differences between PyCharm's native formatter and Ruff. Maybe I can configure PyCharm to pick up a Ruff formatting file...

https://docs.astral.sh/ruff/editors/

File: thermalnetwork/tests/test_base.py (hunk 3 above)
Comment (mitchute) [reply thermalnetwork/tests/test_base.py:55]: I also wasn't clear. In the original, the parentheses seem redundant, but apparently they are doing _something_???

Comment (vtnate) [reply thermalnetwork/tests/test_base.py:55]: There were some redundant parentheses in the original, but at least they weren't stopping it from working. It's cleaned up now.

**Comment by JingWang-CUB** [comment]
Creating a clean conda environment with Python 3.9 worked. All tests passed.
============================= test session starts =============================
platform win32 -- Python 3.9.19, pytest-8.3.2, pluggy-1.5.0
rootdir: C:\GitHub\ThermalNetwork
configfile: pyproject.toml
testpaths: thermalnetwork/tests
plugins: anyio-4.4.0, cov-4.1.0
collected 6 items

thermalnetwork\tests\test_ground_heat_exchanger.py .                     [ 16%]
thermalnetwork\tests\test_heat_pump.py .                                 [ 33%]
thermalnetwork\tests\test_network.py ....                                [100%]

---------- coverage: platform win32, python 3.9.19-final-0 -----------
Coverage HTML written to dir htmlcov


======================== 6 passed in 216.90s (0:03:36) ========================

**Comment by mitchute** [comment]
@vtnate Anything else to do here?

**Comment by vtnate** [comment]
@mitchute This is great! Perhaps remove testing on macos in CI? I think it's good to go. I'll close the other PR.

**Comment by mitchute** [comment]
> Perhaps remove testing on macos in CI?

Yeah, up to you on that one. I'm good either way.

**Pull request merged**